Equilibrium refactor and enhancements

LiamPattinson opened the pull request:
This PR is a follow-up from PR https://github.com/pyro-kinetics/pyrokinetics/pull/128. It's mostly a rebase of that PR on top of the latest Pyrokinetics changes, so please see the wall-of-text I wrote there for more info. For info on how to use these new features, I'd recommend checking out the tutorial I've written in `docs/tutorials/equilibrium.rst`.

Summary of work from PR https://github.com/pyro-kinetics/pyrokinetics/pull/128:

- `Equilibrium` now takes a fixed set of parameters to its `__init__` function. All parameters provided are scalars or arrays.
- `Equilibrium` can be initialised without reading a file. Building an `Equilibrium` from a file is now delegated to the function `read_equilibrium`.
- `Equilibrium` acts like an Xarray dataset, with some variables defined on the $\psi$ grid and others on the $(R,Z)$ grid.
- Using [pint-xarray](https://pint-xarray.readthedocs.io/en/stable/), all variables have units. Units are converted when passing arrays into `Equilibrium.__init__`.
- The function `flux_surface` returns a new class instead of just the $(R,Z)$ path of the flux surface. The `FluxSurface` class contains all of the information needed to produce a `LocalGeometry` from a global `Equilibrium` (arguably it contains too much information, and could be stripped down a bit).
- Refactor of GEQDSK and TRANSP readers to use the new system.
- Cleaner method to register `Reader` classes.
- Added a lot to the docs.
- Units are discarded from `Equilibrium`/`FluxSurface` on conversion to `LocalGeometry`, as that section of the code will need its own refactor to add units.

Changes from PR https://github.com/pyro-kinetics/pyrokinetics/pull/128:

- Fixed `DatasetWrapper` so that global attributes can have units.
- `r` and `z` coordinates are now `R` and `Z` throughout. Similarly, `br` is back to `b_radial`, `bp` is `b_poloidal`, etc.
- Can save `Equilibrium` objects using `eq.to_netcdf("myfile.nc")`, and read them back in using `eq = read_equilibrium("myfile.nc")`, so Pyrokinetics itself is recognised as an equilibrium file type.
- `FluxSurface` derivatives are generated on-the-fly using attribute lookup, i.e `fs.dq_drho` gives the derivative of the safety factor with respect to $\rho$, `fs.dR_major_dr_minor` gives the Shafranov shift.
- Split `normalisations.py` into two files, as the dependency on `LocalGeometry` was causing circular import errors (normalisations depends on LocalGeometry, which depends on Equilibrium, which depends on normalisations in order to use `ureg`). There's probably a cleaner fix for this out there, such as removing the explicit dependency on `LocalGeometry` and just passing `a_minor` instead. For now, I moved everything up to `PyroUnitRegistry` into `units.py`.
- Added a tutorial section to the docs, explaining how to use `Equilibrium` and `FluxSurface` objects.
- Added plotting functions to `Equilibrium` and `FluxSurface`.

The plotting functions might not produce anything of publication quality, but you can pull of some cool things pretty easily:

```python
import pyrokinetics as pk
eq = pk.read_equilibrium(pk.eq_templates["GEQDSK"])
ax = eq.contour(levels=50)
ax = eq.flux_surface(0.9).plot_path(ax=ax)
```
![testplot](https://user-images.githubusercontent.com/14332134/207939239-5e3a86cf-bf7b-469a-8bce-f5632f8b7864.png)

Let me know if you'd like to see any further changes, or if there's some aspect of the design you think should be changed.

LiamPattinson committed:
Refactoring Equilibrium

- Now has a consistent __init__ function with strict requirements.
- Spline fitting is now a hidden implementation detail. Each field
  on the RZ or psi grid is accessed via a function.
- Implemented using a new generic 'DatasetWrapper' class, which
  stores data in an Xarray Dataset and provides utility functions
  such as __getitem__ and __getattr__. The benefits over using a
  raw Dataset is that the user can set a dedicated __init__, and we
  can set our own metadata. The user may write these classes to disk.
- Implemented FluxSurface class, which is expected to be used as
  an interface between Equilibrium and LocalGeometry.
- Factory/Reader bits have been simplified. Instead of registering
  in the __init__.py file, a decorator method is applied as the
  class is defined. Users can also use a method 'read_equilibrium'
  instead of dealing with factory classes directly.
- Aiming for a more 'Pythonic' naming system. Module names are no
  longer simply the name of the main class or function, and are
  written in lower case. Variable names are also now lower case,
  though this may lead to confusion with regards to conventions such
  as 'R = major' radius and 'r = minor radius'

TODO

- Should use pint to enforce units.
- LocalGeometry will need to be updated similarly, and changes
  should be pushed throughout the code.
- TRANSP needs to be converted to the new system.
- Docs are well developed, but still a work in progress.
- A short tutorial should be written and integrated into the docs.
- TESTS!

LiamPattinson committed:
Removed __getattr__ from DatasetWrapper

This exposed too many Xarray Dataset functions to the derived
class. Opted to instead define simple properties for each attr
in Equilibrium.

LiamPattinson committed:
Updated TRANSP eq reader to new methods.

LiamPattinson committed:
WIP adding units to equilibrium

LiamPattinson committed:
Added units to Equilibrium, GEQDSK, TRANSP

Now makes extensive use of pint and pint-xarray to manage the
units used by Equilibrium and associated classes. This handles
issues like TRANSP working in cm instead of m seamlessly, and
going forwards will it easier to normalise data for each
gk code.

TODO
----

- Add units to FluxSurface
- Push changes to LocalGeometry, remove older Equilibrium bits
- T E S T S

LiamPattinson committed:
Added units to FluxSurface

LiamPattinson committed:
Fixed bugs related to force_ndarray for units

monkeypatching f90nml Namelists to convert numpy scalars to their
underlying types

LiamPattinson committed:
Rebased equilibrium update

LiamPattinson committed:
Naming updates, new features, bug fixes

- bug fix to FluxSurface
- DatasetWrapper now redirects 'coords', 'data_vars', 'attrs'
  and 'dims' to the underlying dataset.
- DatasetWrapper records session start as well as time of creation.
- Renamed r,z -> R,Z. Magnetic field functions/quantities br, bz,
  bp, bt expanded to b_radial, b_vertical, b_poloidal, b_toroidal

LiamPattinson committed:
Replaced Equilibrium with new system, expanded tests

- Changes made to Pyro, LocalGeometry classes to use new Equilibrium
- DatasetWrapper now has units for global attributes
- FluxSurface derivatives improved (might be slowing things down?)
- Greatly expanded tests for Equilibrium

LiamPattinson committed:
DatasetWrapper __getattr__ tries super() method first

LiamPattinson committed:
Updated docs for new Equilibrium

LiamPattinson committed:
Added plot function to Equilibrium, black/flake8 fixes

LiamPattinson committed:
Equilibrium at top-level __init__, better DatasetWrapper repr

LiamPattinson committed:
Added contour plot to Equilibrium, tests for Equilibrium.plot

LiamPattinson committed:
Added flux surface plotting functions, further plotting tests

LiamPattinson committed:
Started work on Equilibrium tutorial

LiamPattinson committed:
Fixed DatasetWrapper attribute units bug when reading/writing netcdf

Also added test to ensure Equilibrium objects survive the
roundtrip of read/write.

LiamPattinson committed:
Added EquilibriumReaderPyro

Can now use read_equilibrium() on Pyro generated equilibrium files.

LiamPattinson committed:
Finished tutorial for Equilibrium/FluxSurface

bpatel2107 commented on pyrokinetics/equilibrium/geqdsk.py:
```diff
@@ -0,0 +1,128 @@
+from ..units import ureg as units
+
+import numpy as np
+from freegs import _geqdsk
```
Might be worth switching over to reading via FreeQDSK which is a bespoke GEQDSK file reader rather than using a GEQDSK reader in a equilibrium solver. I think all the variable names are the same so hopefully its a simple switch


https://github.com/ZedThree/FreeQDSK

bpatel2107 commented on pyrokinetics/examples/example_JETTO.py:
```diff
@@ -14,7 +14,7 @@ def main(base_path: Union[os.PathLike, str] = "."):
     # Load up pyro object
     pyro = Pyro()
 
-    pyro.load_global_eq(eq_file=eq_file, eq_type="GEQDSK", psi_n_lcfs=1.0)
```
The point of `psi_n_lcfs` was to artificially change where the actual LCFS is. The reason for this was that sometimes, when doing the contour fittings, if the actual value of `psi` was used to specify the LCFS, a connected contour wouldn't be found. So I gave the option of slightly shifting the value of `psi` for the LCFS.

Admittedly this is more of a issue with the contour fitting rather than the equilibrium but it was a simple workaround. I see that you can now you specify `psi_lcfs` directly. Can we add a multiplier to slightly change the value if needed. If you are loading in an equilibrium you likely won't know what `psi_lcfs` is but you could easily specify `psi_n_lcfs` which defines the `psi` used

bpatel2107 commented:
So having a quick look it seems like `psi_n_lcfs` is actually in here but implementation needs a subtle tweak.

The only slight issue is that in the following lines the interpolation assumes the final element in the GEQDSK arrays matches `psi_n_lcfs` but that is always at `psi_n = 1.0` so if we were to use `psi_n_lcfs=0.9` then it would be slightly off.

https://github.com/PlasmaFAIR/pyrokinetics/blob/unstable/pyrokinetics/equilibrium/EquilibriumReaderGEQDSK.py#L49-L52

What we should do is to fit that using the full range and then redefine the `psi_n` array using `psi_n_lcfs` after `q`, `f_psi`, `ff'` and `pressure` are set.

Are you able to look at the merge conflicts here?

LiamPattinson commented:
@bpatel2107 I'll be taking a look at this for a few hours today, and I'll probably continue working on it tomorrow and on Friday. I'll try to get `psi_n_lcfs` working again, and I'll see what's going wrong with the merge conflicts. I got some further feedback in person from @ZedThree, so I'll be making those changes too.

LiamPattinson committed:
Merge branch 'unstable' into enhancements/equilibrium_v2

LiamPattinson committed:
Reintroduced psi_n_lcfs to geqdsk reader

LiamPattinson committed:
Fixed interpretation of psi_n_lcfs

LiamPattinson committed:
Also adjust a_minor when psi_n_lcfs is not 1

LiamPattinson commented:
Sorry, disregard the last few commits, I don't think I understand how `psi_n_lcfs` is supposed to be used.

In the original code, this changes the grid on which each quantity is defined, and therefore adjusts the derivatives:

https://github.com/pyro-kinetics/pyrokinetics/blob/95176763ea819eb7abf265c064961a47cf9067fa/pyrokinetics/equilibrium/EquilibriumReaderGEQDSK.py#L42
https://github.com/pyro-kinetics/pyrokinetics/blob/95176763ea819eb7abf265c064961a47cf9067fa/pyrokinetics/equilibrium/EquilibriumReaderGEQDSK.py#L49-L53

This also means that if we set `psi_n_lcfs` to 0.9, and if the user tries to get data at `psi_n=0.95` later, it'll be determined by extrapolating from the splines. However, `a_minor` is set by determining the maximum width of the last flux surface:

https://github.com/pyro-kinetics/pyrokinetics/blob/95176763ea819eb7abf265c064961a47cf9067fa/pyrokinetics/equilibrium/EquilibriumReaderGEQDSK.py#L69-L75

https://github.com/pyro-kinetics/pyrokinetics/blob/95176763ea819eb7abf265c064961a47cf9067fa/pyrokinetics/equilibrium/EquilibriumReaderGEQDSK.py#L82

This isn't affected by `psi_n_lcfs` as far as I can tell, so does this imply that `psi_n` and `rho` are normalised with respect to different flux surfaces? My reading is that `psi_n` is normalised with respect to a flux surface outside the bounds of the GEQDSK file, while `rho` is normalised with respect to the GEQDSK LCFS.

Is this the expected behaviour? My current implementation does the opposite, with everything calculated on the `psi` grid as usual (not `psi_n`), and if the user sets `psi_n_lcfs=0.9`, it sets `psi_lcfs` in the Equilibrium object to `0.9 * psi_lcfs + 0.1 * psi_axis`, and calculates `a_minor` on the same flux surface:

```python
        # Adjust normalising factors if psi_n_lcfs is not 1.0
        psi_lcfs = psi_n_lcfs * psi_lcfs + (1.0 - psi_n_lcfs) * psi_axis
        a_minor = UnitSpline(psi_grid, r_minor)(psi_lcfs)
```

LiamPattinson committed:
Better UnitSplines

Uses SciPy splines by composition, not inheritance

LiamPattinson committed:
Minor optimisation on psi_n_lcfs

LiamPattinson committed:
Equilibrium refactor

- reader factory is now a class-level attribute instead of being
  in global scope: _equilibrium_readers -> Equilibrium._readers
- equilibrium_reader decorator refactored to be a classmethod on
  Equilibrium, Equilibrium.reader. The plain function version
  now calls this.
- read_equilibrium refactored to be a classmethod on Equilibrium,
  Equilibrium.from_file. The plain function version now calls
  this.
- supported_equilibrium_types refactored to be a classmethod on
  Equilibrium, Equilibrium.supported_types. The plain function
  version now calls this.
- asanyarray(..., dtype=float) -> asfarray(...).
- Added descriptive comments in places

LiamPattinson committed:
Fix docs typos

LiamPattinson committed:
Added option to set equal aspect ratio on plots

LiamPattinson committed:
Removed dynamic derivatives from FluxSurface, constrained attrs

FluxSurface no longer allows the user to determine derivatives
  dynamically.
FluxSurface now requires a smaller selection of quantities as input
  arguments, based on those in the OMAS schema gyrokinetics.flux_surface
  It now needs to be initalised with things such as magnetic shear,
  pressure gradient (wrt r_minor), shafranov shift, etc.
It may be possible to remove some further variables in future:
  psi_gradient, f, p?

LiamPattinson commented:
I've made some changes to the PR, but it's nothing major:

- A bunch of things that were at global scope are now class-level attributes/classmethods on `Equilibrium`. Plain function alternatives are still available, but now these simply forward args to their corresponding classmethods. Now we can build an `Equilibrium` from a file using either `Equilibrium.from_file` or `read_equilibrium`.
- `FluxSurface` has been cleaned up, and now its attributes are just the things needed by `LocalGeometry` classes, though named following the [OMAS schema `gyrokinetics.flux_surface`](https://gafusion.github.io/omas/schema/schema_gyrokinetics.html). I think it may be possible to merge these classes with some further work, as they're both describing the same thing -- the only fundamental difference is that `LocalGeometry` is parameterised while `FluxSurface` isn't. That's something for a future PR though.
- Minor updates throughout

I think I'm following RE `psi_n_lcfs`, and if I understand the intent correctly, then I think my current implementation does what's required. Since all of the `psi` grid quantities are determined on `psi` splines instead of `psi_n` splines, all we're doing is adjusting `psi_lcfs` and `a_minor` slightly, and nothing else is changed in `Equilibrium`. The only practical difference is that if the user sets `psi_n_lcfs=0.9`, and then looks up something at `psi_n=0.5`, it'll actually end up looking at `psi_n=0.45` in the GEQDSK file. Let me know if it looks like this won't achieve what's needed or if I'm still misunderstanding.

I'll see if I can get a switch to FreeQDSK working tomorrow, though that might require some quick work on FreeQDSK itself.

bpatel2107 commented:
I believe this will still fail for certain GEQDSK files as the error occurs when finding all of the flux surfaces which happens before your tweak here.

If `psi_n_lcfs` is not specified then you are still trying to fit to the actual LCFS here as `psi_grid` has `psi_lcfs` in it.


Essentially we could  do this line first

https://github.com/pyro-kinetics/pyrokinetics/blob/89a47c50c6d7a0c762a01a9724ca0d83978d880f/pyrokinetics/equilibrium/geqdsk.py#L113

Then adjust the final value in `psi_grid` and then loop through the flux surface fitting here

https://github.com/pyro-kinetics/pyrokinetics/blob/89a47c50c6d7a0c762a01a9724ca0d83978d880f/pyrokinetics/equilibrium/geqdsk.py#L99-L100

Then we could set `a_minor`

https://github.com/pyro-kinetics/pyrokinetics/blob/89a47c50c6d7a0c762a01a9724ca0d83978d880f/pyrokinetics/equilibrium/geqdsk.py#L114

It might be a good idea to set a lower limit on `psi_n_lcfs` to something like 0.99 until we handle this contour issue.

Alternatively we could sidestep this issue and handle this case inside the flux surface fitting. We could have a check if the flux surface fails when `psi_n=1.0` and try again with `psi_n_lcfs=0.999` and raise a warning. Not sure if that is better practice or not.

bpatel2107 commented:
I like the idea of merging `LocalGeometry` and `FluxSurface` as they are describing the same thing

LiamPattinson committed:
Further psi_n_lcfs changes

bpatel2107 commented:
I think what you have done is probably the best we can do. Happy to merge if this is ready

LiamPattinson commented:
I could try to get FreeQDSK in this PR, since the latest FreeGS release isn't compatible with the latest NumPy, but I suppose that could also be a separate PR. Otherwise I don't think I have anything else to add.

LiamPattinson committed:
Added check for reversed psi in GEQDSK files

bpatel2107 commented:
Can add in FreeQDSK in another PR.

We should be careful about the reversing of the psi. It's likely that the GEQDSK there have a different COCOs (coordinate convention https://www.sciencedirect.com/science/article/pii/S0010465512002962) that we need to handle with care. 

For now this is good!
